Unified Build documentation (#11439)
dotnet/arcade: Documentation/UnifiedBuild/Foundational-Concepts.md

# The Unified Build Almanac (TUBA) - Foundational Concepts

## Purpose
This document describes the foundational concepts of Unified Build. The basic principles, the pillars upon which it is built, the goals, etc.

## Introduction

Unified Build for .NET is a combined effort to improve two problematic areas for .NET:
- **True upstream model** – Partners are efficient and can do the same things Microsoft can.
- **Product build and servicing complexity** – Taming .NET's incredibly complex infrastructure.

.NET is a collection of repositories that collectively ship together to form a coherent product. Each individual repository in the graph contributes build outputs that are consumed in other repositories, and potentially shipped to customers. Therefore, a 'build' of .NET is not just a build of a single repository. .NET maintains two ways of building this product. 
- **Source-build** - Building an *aggregate* of all of the repositories on a single machine (no orchestration) in a single configuration to produce a .NET SDK that is shippable by our Linux source-build partners.

> - **Microsoft's official build** - The process of building each node in the dependency graph, updating any downstream with newly produced outputs, testing and committing those updates, and then repeating until all changes have flowed through the graph to all possible downstream nodes. As you might imagine, this process is complex. New changes in repositories reset the build. Coordinating branch management for more than a dozen repositories across multiple releases of .NET is challenge. There is a high degree of infrastructural overhead. Furthermore, even within repository builds there tends to be a large degree of complexity and orchestration.

Reviewer: > any downstream

"repo"?

> Maintaining both these methodologies is expensive. There are meaningful differences between how source-build and Microsoft's official build produce binaries. There are different restrictions on what inputs are allowed to be used in source-build as compared to Microsoft's build. Even more, the source-build methodology today does not support building the full breadth of assets that Microsoft ships on a monthly cadence. From a community perspective, these differences contribute to a lack of approachability and an inability to have a meaningful upstream/downstream model for collaboration.

Reviewer: > the source-build methodology today does not support building the full breadth of assets that Microsoft ships

An example seems warranted. Otherwise it seems a bit scary. Saying "Windows" would help.

One insight is that Microsoft's build, source-build, and meaningful upstream/downstream collaboration are **overlapping problems**, and can be satisfied by the same solution if considered together. Our whole-product repo structure is very costly for us and unapproachable by others. We have spent a lot on these topics over the years, with hard-won progress, and no end in sight. A different repo design-point could provide us with a better and cheaper system.

> **Proposition** Focusing on the source-build philosophy (aggregate code base is the source-of-truth) as the way that the community **and** Microsoft produce the product will drastically development operations, servicing, and community interaction.

Reviewer: ```suggestion
**Proposition** Focusing on the source-build philosophy (aggregate code base is the source-of-truth) as the way that the community **and** Microsoft produce the product will drastically improve development operations, servicing, and community interaction.
```

## Unified Build goals

The following are goals and that will be met by the Unified Build project. Please note that ".NET distro maintainers" includes Microsoft.

> - .NET adopts industry standard upstream/downstream practices.

Author: Who participates is about the VMR. I'll rephrase this though.

> - .NET adopts industry standard upstream/downstream practices.
> - A single git commit denotes all product source for a particular .NET build. All commits are coherent
> - A single repo commit can produce a shippable build
> - .NET's build shall be able to create a specific platform's distribution in a single build environment.
> - .NET distro maintainers shall be able to efficiently update and build .NET (both collaboratively and separately) through the entire lifecycle of a .NET version (first to last commit).
> - .NET distro maintainers can produce downstream distributions without use of Microsoft provided services.
> - .NET distro maintainers shall be able to meet provenance and build environment requirements for their distributions.
> - .NET distro maintainers shall be able to coordinate patching of downstream distributions.
> - .NET distro maintainers can run verification tests against the built product.
> - .NET developers shall be able to easily produce full product builds for testing, experimentation, etc.
> - .NET developers shall be able to work efficiently on the section of the product for which they are concerned

Reviewer 2: ```suggestion
- .NET adopts industry standard upstream/downstream practices.
- A single git commit denotes all product source for a particular .NET build. All commits are coherent.
- A single repo commit can produce a shippable build.
- .NET's build shall be able to create a specific platform's distribution in a single build environment.
- .NET distro maintainers shall be able to efficiently update and build .NET (both collaboratively and separately) through the entire lifecycle of a .NET version (first to last commit).
- .NET distro maintainers can produce downstream distributions without use of Microsoft provided services.
- .NET distro maintainers shall be able to meet provenance and build environment requirements for their distributions.
- .NET distro maintainers shall be able to coordinate patching of downstream distributions.
- .NET distro maintainers can run verification tests against the built product.
- .NET developers shall be able to easily produce full product builds for testing, experimentation, etc.
- .NET developers shall be able to work efficiently on the section of the product for which they are concerned.
```

### **Upstream partners primarily focus on the VMR structure**

The VMR focused build enables partners to build and test .NET more easily. This should enable them to more effectively contribute when desired. While they of course may contribute to either the VMR or individual repos, it's expected that they will choose to work in the VMR.

### **Individual developers continue to work primarily in individual repositories for mainline development**

Though they may work in the VMR, it is expected that many developers will continue to primarily work in the individual repositories for mainline development, especially those developers who focus on specfic areas of .NET. Maintaining the individual repository workflows enables high-velocity development for changes in isolated components. It also allows for continued community building in specific product areas. Cross-repo breaking changes will be resolved in the VMR, then build outputs and additional source changes will flow back into the individual repositories.

> 

Author: They will stay the same. Maestro has no effect on developer builds today, and Arcade will not change.

> 

Author: And for the product build, Maestro will still be active for some flows.

## Rules of Unified Build

Unified Build and the associated infrastructure are built around the rules outlined below, with their associated reasoning. There are two motivations behind these rules: 

> - **Enabling non-Microsoft .NET distro maintainers to succeed** - Microsoft is the primary maintainer of the .NET upstream. Our choices about infrastructure technologies and build methodologies have real effects on our downstream partners, like RedHat, Canonical, or Unity. If we make poor choices, we make it harder or even impossible for our partners to be successful in delivering the product to their customers. While Microsoft has the vast engineering resources to be able to build, service and deliver to all sorts of customers in all manner of scenarios, this is not without cost. To grow our product the way we want, we need to enable our partners to serve their customers without handholding.

Reviewer: ```suggestion
- **Enabling any .NET distro maintainer to succeed** - Microsoft is the primary maintainer of the .NET upstream. Our choices about infrastructure technologies and build methodologies have real effects on our downstream partners, like RedHat, Canonical, or Unity. If we make poor choices, we make it harder or even impossible for our partners to be successful in delivering the product to their customers. While Microsoft has the vast engineering resources to be able to build, service and deliver to all sorts of customers in all manner of scenarios, this is not without cost. To grow our product the way we want, we need to enable our partners to serve their customers without handholding.
```

While our partners may not always build what they release based on the upstream, as they may use private pre-release channels to obtain sources of upcoming versions, those releases must eventually be made open-source in the upstream to maintain their open source status.

***Note: What Microsoft builds and releases for Linux, MacOS or any other platform, is not intended to be ‘bit for bit’ identical with what a partner may be able to build using the upstream. Signing, for example, results in a slightly different product. The product shall be functionally identical***

### **The build shall not require any orchestration to build artifacts for a specific platform distribution**

*This rule focuses on simplicity for .NET distro maintainers, including Microsoft, and ensuring that .NET distro maintainers can meet provenance and build environment requirements.*

Orchestration is most associated with CI systems, though other systems like Maestro also provide orchestration in some forms. Orchestration, while useful, tends to:
- Encourage complexity
- Add fragility to the overall build process because of an increased use of distributed services.

> - Encourage multi-machine workflows, which are incompatible with many organization's build requirements.

Reviewer: ```suggestion
- Encourage multi-machine workflows, which are incompatible with many organizations' build requirements.
```

.NET has used a multi-repository build through .NET 7. Through a combination of individual repository builds, dependency flow and tracking, the shipping artifacts for a given release are produced. In the release process, the dependency tracking system (Maestro) traverses the graph of individual repository builds contributing to a given release, locating and downloading all the artifacts marked as shipping. These artifacts are then modified and shipping.

This multi-build ‘gathering’ process is not allowed in Unified Build. The VMR build must produce all artifacts that are to ship. If gathering were to be allowed, this would imply that orchestration could be required to build product (dependency flow through multiple repos), and that external services (Maestro) would be required to track the artifacts in more than one repo build.
*Note: This is not to say that Maestro cannot be used in some capacity for VMR operation. It just cannot be required. An allowable use may be to flow non-stable, non-shipping assets referenced by the VMR, like test or optimization data.*

### **There may be no pre-release, non-final stable binary flow**

*This rule focuses on simplicity and robustness.*

> Stable binary flow is defined as any flow of binaries from a build to a repository, where the identity of those binaries does not vary from build to build with respect to their packaging systems **and** binaries with the same identity are produced *more than once*. This flow is commonly used in RTM/Servicing releases of .NET. Stable binary flow has been extremely costly to implement and makes the shipping product harder to understand. The basic problems are:

Reviewer: Very long first sentence. Is a "when" missing around "with the same ..."?

Author: Broke this up
